internal/store/generic.go

```diff
@@ -0,0 +1,405 @@
+		if err != nil {
+			return nil, false, false, err
+		}
+
+		err = validationCallback(existingAPIResource, modifiedAPIResource)
+		if err != nil {
+			return nil, false, false, err
+		}
+	}
+
+	if exists && mode == ModeCreateOnly {
+		return nil, false, false, flterrors.ErrDuplicateName
+	}
+	if !exists && mode == ModeUpdateOnly {
+		return nil, false, false, flterrors.ErrResourceNotFound
+	}
+
+	s.IntegrationTestCreateOrUpdateCallback()
+
+	var retry bool
+	if !exists {
+		retry, err = s.createResource(ctx, model)
+	} else {
+		retry, err = s.updateResource(ctx, fromAPI, existing, model, fieldsToUnset)
+	}
+	if err != nil {
+		return nil, false, retry, err
+	}
+
+	if callback != nil {
+		modifiedAPIResource, err := s.modelPtrToAPI(model)
+		if err != nil {
+			return nil, false, false, err
+		}
+
+		callback(orgId, existingAPIResource, modifiedAPIResource)
+	}
+
+	apiResource, err := s.modelPtrToAPI(model)
+
+	return apiResource, !exists || P(existing).HasNilSpec(), false, err
+}
+
+func (s *GenericStore[P, M, A, AL]) getExistingResource(ctx context.Context, name string, orgId uuid.UUID) (*M, error) {
+	var existingResource M
+	if err := s.db.WithContext(ctx).Where("name = ? and org_id = ?", name, orgId).First(&existingResource).Error; err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, nil
+		}
+		return nil, ErrorFromGormError(err)
+	}
+	return &existingResource, nil
+}
+
+func (s *GenericStore[P, M, A, AL]) createResource(ctx context.Context, resource P) (bool, error) {
+	resource.SetGeneration(util.Int64ToPtr(1))
+	resource.SetResourceVersion(util.Int64ToPtr(1))
+
+	result := s.db.WithContext(ctx).Create(resource)
+	if result.Error != nil {
+		err := ErrorFromGormError(result.Error)
+		return err == flterrors.ErrDuplicateName, err
+	}
+	return false, nil
+}
+
+func (s *GenericStore[P, M, A, AL]) updateResource(ctx context.Context, fromAPI bool, existing, resource P, fieldsToUnset []string) (bool, error) {
+	if fromAPI && (util.DefaultIfNil(resource.GetOwner(), "<NIL>") != util.DefaultIfNil(existing.GetOwner(), "<NIL>")) {
+		return false, flterrors.ErrUpdatingResourceWithOwnerNotAllowed
+	}
+
+	sameSpec := resource.HasSameSpecAs(existing)
+	if !sameSpec {
+		if fromAPI {
+			hasOwner := len(lo.FromPtr(existing.GetOwner())) != 0
+			if hasOwner {
+				// Don't let the user update the spec if it has an owner
+				return false, flterrors.ErrUpdatingResourceWithOwnerNotAllowed
+			}
+		}
+
+		// Update the generation if the spec was updated
+		resource.SetGeneration(lo.ToPtr(lo.FromPtr(existing.GetGeneration()) + 1))
+	}
+
+	if resource.GetResourceVersion() != nil &&
+		lo.FromPtr(existing.GetResourceVersion()) != lo.FromPtr(resource.GetResourceVersion()) {
+		return false, flterrors.ErrResourceVersionConflict
+	}
+
+	resource.SetResourceVersion(lo.ToPtr(lo.FromPtr(existing.GetResourceVersion()) + 1))
+
+	selectFields := []string{"spec"}
+	if resource.GetKind() == api.DeviceKind {
```

Comment:
Can we avoid adding specific resource code in the generic store ?

Reply:
I can add a method to each resource to get additional fields?  Not sure if it's worth it